javatests/arcs/core/data/TypeUtilsTest.kt

```diff
@@ -0,0 +1,175 @@
+/*
+ * Copyright 2020 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+package arcs.core.data
+
+import arcs.core.data.expression.InferredType
+import arcs.core.data.expression.MapScope
+import com.google.common.truth.Truth.assertThat
+import kotlin.test.assertFailsWith
+import org.junit.Before
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.junit.runners.JUnit4
+
+private const val testField = "test"
+
+/** Tests for TypeUtils. */
+@RunWith(JUnit4::class)
+class TypeUtilsTest {
+
+  @Before
+  fun setUp() {
+    SchemaRegistry.clearForTest()
+  }
+
+  private fun makeSchemaForType(fieldType: FieldType, isCollection: Boolean) =
+    EntityType(
+      Schema(
+        setOf(SchemaName("Test")),
+        SchemaFields(mapOf(testField to fieldType), emptyMap()),
+        hash = "42"
+      ).also {
+        SchemaRegistry.register(it)
+      }
+    ).let {
+      if (isCollection) CollectionType(it) else SingletonType(it)
+    }
+
+  val allTypes = mapOf(
+    FieldType.Boolean to InferredType.Primitive.BooleanType,
+    FieldType.Byte to InferredType.Primitive.ByteType,
+    FieldType.Char to InferredType.Primitive.ShortType,
+    FieldType.Short to InferredType.Primitive.ShortType,
+    FieldType.Int to InferredType.Primitive.IntType,
+    FieldType.Instant to InferredType.Primitive.LongType,
+    FieldType.Long to InferredType.Primitive.LongType,
+    FieldType.BigInt to InferredType.Primitive.BigIntType,
+    FieldType.Double to InferredType.Primitive.DoubleType,
+    FieldType.Float to InferredType.Primitive.FloatType,
+    FieldType.Number to InferredType.Primitive.NumberType,
+    FieldType.Text to InferredType.Primitive.TextType
+  )
+
+  private fun schemaFieldOf(type: InferredType): InferredType = when (type) {
+    is InferredType.ScopeType -> type.scope.lookup(testField)
+    is InferredType.SeqType -> schemaFieldOf(type.type)
+    else -> throw IllegalArgumentException()
+  }
+
+  private fun mapFieldType(fieldType: FieldType, isCollection: Boolean = false) =
+    schemaFieldOf(mapTypeToInferredType(makeSchemaForType(fieldType, isCollection)))
```

Comment:
This is mind bending a bit :D

Reply:
1) create a Schema class with the FieldType, wrap is as EntityType, and then SingletonType/CollectionType
2) map it to an InferredType (ScopeType)
3) Extract the Scope from ScopeType, and look up the InferredType of the field. I can copy this comment to KDoc.